Add support for physical qubits

jlapeyre opened the pull request:
### Example

The most basic example is
```python
from qiskit_qasm3_import import parse

program = """
OPENQASM 3.0;

reset $2;
"""

circuit = parse(program)
circuit.draw()
print(circuit._layout)
```
which prints
```
0 ─|0>─       

TranspileLayout(initial_layout=Layout({
2: Qubit(QuantumRegister(1, 'qr'), 0)
}), input_qubit_mapping={Qubit(QuantumRegister(1, 'qr'), 0): 2})
```

### Tasks

- [x] Refactor non-DRY parts
- [x] Tests
- [x] Raise some errors, for example when mixing types of qubits
- [x] Doc strings [No need for doc strings]
- [x] Release note

### Additional Tasks

- [x] Rewrite `AddressingMode`
- [x] Split some tests into separate functions
- [x] Test (and implement) rejecting references to hardware qubits in function body.
- [x] Test hardware qubits in nested scopes.
- [x]  [GJL] Test that layout is correctly generated for hardware qubits.
- [x]   [GJL] Test hardware qubit first seen in local scope broken.
- [x] Remove commented out cruft in OQ code.
- [x] Second arg to `resolve_condition` should not be optional
- [x] Second arg to `resolve_condition` should be of one type (`State`)
- [x] Create and use a type for hardware qubits, eg `type.HardwareQubit`.
- [x] In `ValueResolver.visit` no need to handle missing context explicitly.
- [x] Check entry point `resolve` probably shouldn't accept `None` as an input to `context` here, since it's now required.
- [x] Either pass all state data in `resolve` or store in class, but don't mix. See https://github.com/Qiskit/qiskit-qasm3-import/pull/2#discussion_r1088943129
- [x] In `ConvertVisitor.convert` fix inefficiency in iterating through keys.
- [x] Make `protected-access` lint exception more locally scoped.
- [x]  Disable too many lines lint check globally
- [x]  AddressingMode: use enum for the mode number. Or switch to using just an enum
- [x] Add `HardwareQubit` to the docs/index.rst
- [x] Remove dead code, old comments, etc.
- [ ] test that the `AddressingMode` is being set correctly and propagated all the way up to the global scope
- [x] add a test that we can use hardware addressing in the main program, but still have user-defined gates
- [x] fairly sure Qiskit says that backends are required to label their hardware qubits starting from 0 with no gaps
- [ ] add an option to the top-level convert function to allow the user to specify the total size of the backend
- [x] Python `Enum` instances can (should) be compared by is checks
- [x] the triple-nested if/else now, can we refactor this a little to do early return
- [x] Rewrite `SymbolTable` with stack-like structure.
- [x] test that hardware qubits can be used outside the local scope they're first seen in
- [x] test that "hardware mode" being set in a local scope and then referencing virtual qubit after that scope is exited causes an error to be raised.
- [x] Remove test involving identifer `$q`.
- [x] Remove `hardware_qubits` method. use a different design.

### Notes

EDIT: Regarding `_layout` and `TranspileLayout` below: PR https://github.com/Qiskit/qiskit-terra/pull/9486 now provides an interface and some documentation.

`QuantumCircuit` has an undocumented (including in comments) property `_layout`. It is initialized to `None`. It seems that sometimes it is set to a `Layout` and sometimes a `TranspileLayout`. I find that printing/drawing the circuit fails if `_layout` is a `Layout`, because the output methods reference `_layout.initial_layout`, which does not exist.  So I use `TranspileLayout`, which does have `_layout.initial_layout`.

The hardware qubits are referred to by numbers. These are mapped `Qubit`s in a  `QuantumRegister` , `"qr"`, in the order that they were seen in the OQ3 code; the first is mapped to `qr[0]` , and so on.

The references to hardware qubits are collected in `ValueResolver.visit_Identifier`. The layout is then created and attached to the circuit as the last step in the method `ConvertVisitor.convert`.

Here is [the code that adds the layout](
https://github.com/jlapeyre/qiskit-qasm3-import/blob/d6d30936aaf0604161664827ac07a9b4f6f5f6c7/src/qiskit_qasm3_import/converter.py#L203-L211)
```python
            names = filter(is_physical, symbols.keys())
            intlist = physical_qubit_identifiers_to_ints(names)
            qr = QuantumRegister(len(intlist), "qr")
            initial_layout = Layout.from_intlist(intlist, qr)
            input_qubit_mapping = dict(zip(qr, intlist))
            layout = TranspileLayout(initial_layout, input_qubit_mapping)
            state.circuit._layout = layout
```

Previously, the method `ValueResolver.resolve` takes only the node to be resolved and a symbol table. I added as a third argument the context (an instance of `State`) because it contains the circuit, and each physical qubit must be added to the circuit as are the virtual qubits in `ConvertVisitor.visit_QubitDeclaration`.

Closes #1

Closes the proxy issue https://github.com/Qiskit/qiskit-terra/issues/9435

jlapeyre commented:
`pylint -sn -rn src tests` is failing in CI. This command failed to run locally under Python 3.11 because I had `wrapt` v 1.12 installed in my virtual env. After upgrading to `wrapt` v 1.14, the command runs correctly.

Perhaps this requirement should be added somewhere. Presumably not in this repo, because `wrapt` is an indirect dependency. If this has actually been fixed by a version requirement in the direct dependency that is using `wrapt`, we could make a requirement here. I have not tracked this down.

jakelishman commented:
It's transitive and already fixed - not our responsibility, and we'd just be polluting our repo if we did it.

jakelishman commented:
I provided a `tox` configuration, so you can run `tox -e py311` to run the tests under Python 3.11, or `tox -e lint` to run the lint tests in an environment isolated from your current one, which should alleviate a lot of out-of-date environment issues.

jlapeyre commented:
Running `tox -e lint` gives me the following
```
************* Module qiskit_qasm3_import.types
src/qiskit_qasm3_import/types.py:5:11: E1101: Module 'abc' has no 'ABC' member (no-member)
src/qiskit_qasm3_import/types.py:16:5: E1101: Module 'abc' has no 'abstractmethod' member (no-member)
src/qiskit_qasm3_import/types.py:58:4: W0231: __init__ method from base class 'Type' is not called (super-init-not-called)
```

Looking in `.tox/lint/bin` shows that this is using 3.11

jakelishman commented:
That's pretty weird - looks like a bug in the pylint 2.14 series, because `abc.abstractmethod` and `abc.ABC` are defined as normal Python objects in the standard library.  There isn't even any weird C-extension re-exports for those.  I'll bump the pylint version to 2.15.x, since that has no errors with Python 3.11.

Fwiw, all the non-generated `tox` targets use the "system" `python` binary.  It's not super ideal, but if you have a Python 3.10 environment that also has `tox` installed, you can do `/path/to/python3.10 -m tox [tox options]` (or `/<py310 venv>/bin/tox [tox options]`) to get it to do what you want here.

jlapeyre commented:
@jakelishman , @kdk  This is nearly finished. You may want to take a look at this before I clean up rough edges it to see if we need structural changes to the implementation. In any case, I'll start polishing a bit.

In particular, the section "Notes" in the opening comment above describes the implementation.

jakelishman commented on src/qiskit_qasm3_import/converter.py:
```diff
@@ -183,13 +191,23 @@ class ConvertVisitor(QASMVisitor[State]):
         This is used to generated improved error messages."""
-        return self.visit(node, State(Scope.GLOBAL, source)).circuit
+
+        state = self.visit(node, State(Scope.GLOBAL, source))
+        symbols = state.symbol_table
+        if any(is_physical(name) for name in symbols.keys()):
+            names = filter(is_physical, symbols.keys())
+            intlist = physical_qubit_identifiers_to_ints(names)
+            qr = QuantumRegister(len(intlist), "qr")
+            initial_layout = Layout.from_intlist(intlist, qr)
+            input_qubit_mapping = dict(zip(qr, intlist))
+            state.circuit._layout = TranspileLayout(initial_layout, input_qubit_mapping)
+        return state.circuit
```
This doesn't feel quite right - we're doing ad-hoc type checking within `ConvertVisitor`, and we're not even using the typing information within `Symbol` instance to do the check, we're trying to infer it from the name.  (It's correct for _something_ to infer it from the name in this case, it just should be the type checker, not the converter.)

Instead, I'd add a `HardwareQubit` _type_ to the `types` module, and have the type checker assign all relevant symbols to that type (there's a field in `Symbol` for it).  Then you can iterate through the symbol table just looking for that type, and it keeps the separation of concerns.

jlapeyre replied:
Yes, adding a `HardwareQubit` type here is much better. But I think it's even better to do this in `openqasm3` (as is https://github.com/openqasm/openqasm/pull/441).

jakelishman commented on src/qiskit_qasm3_import/converter.py:
```diff
@@ -183,13 +191,23 @@ class ConvertVisitor(QASMVisitor[State]):
+
+        state = self.visit(node, State(Scope.GLOBAL, source))
+        symbols = state.symbol_table
+        if any(is_physical(name) for name in symbols.keys()):
+            names = filter(is_physical, symbols.keys())
```
Very minor, but this is an inefficiency - we iterate through `symbols.keys()` completely twice.  It'd be a shade cleaner to do
```python
physical_indices = [int(name.name[1:]) for name, value in symbols.items() if isinstance(value.type, types.HardwareQubit)]
if physical_indices:
    qr = QuantumRegister(...)
    ...
```

jlapeyre replied:
Oh yeah. That was one of areas on my refactoring checklist, but it slipped through.

jakelishman commented on src/qiskit_qasm3_import/expression.py:
```diff
@@ -62,43 +76,60 @@ class ValueResolver(QASMVisitor):
     def __init__(self, symbols: Mapping[str, Symbol]):
         self.symbols = symbols
 
-    def resolve(self, node: ast.Expression) -> Tuple[Any, types.Type]:
+    def resolve(self, node: ast.Expression, context=None) -> Tuple[Any, types.Type]:
```
We should either store all the relevant parts of the state in the `ValueResolver` instance, or we should pass them all (including the symbol table) through the `context` parameter.  I don't mind particularly which we choose, but let's be consistent.  If we pass the context as an argument, let's also return it in the tuple in `resolve` at the end, so we're free to mutate the context in whatever manner we like, including replacing the whole thing.

Minor: the entry point `resolve` probably shouldn't accept `None` as an input to `context` here, since it's now required.

jlapeyre replied:
I'm pretty sure the code has been modified to satisfy these items. I'm resolving this to reduce visual clutter.

jakelishman commented on src/qiskit_qasm3_import/expression.py:
```diff
@@ -62,43 +76,60 @@ class ValueResolver(QASMVisitor):
     def __init__(self, symbols: Mapping[str, Symbol]):
         self.symbols = symbols
 
-    def resolve(self, node: ast.Expression) -> Tuple[Any, types.Type]:
+    def resolve(self, node: ast.Expression, context=None) -> Tuple[Any, types.Type]:
         """The entry point to the resolver, resolving the AST node into a 2-tuple of a relevant
         Qiskit type, and the :class:`.Type` that it is an instance of."""
-        return self.visit(node)
+        if context is None:
+            return self.visit(node)
+        return self.visit(node, context)
 
     def visit(self, node: ast.QASMNode, context: None = None) -> Tuple[Any, types.Type]:
-        value, type = super().visit(node, context)
+        if context is None:
+            value, type = super().visit(node)
+        else:
+            value, type = super().visit(node, context)
```
Whatever we do with `None` handling, I don't think there's any need to do this - the super `visit` method will do it.

jakelishman commented on src/qiskit_qasm3_import/expression.py:
```diff
@@ -62,43 +76,60 @@ class ValueResolver(QASMVisitor):
         if name not in self.symbols:
-            raise_from_node(node, f"name '{name}' is not defined in this scope")
-        symbol = self.symbols[name]
+            if is_physical(name):  # Physical qubits are not declared.
+                if not context.addressing_mode.set_physical_mode():
+                    raise_from_node(
+                        node,
+                        "Physical qubit referenced in virtual addressing mode. Mixing modes not currently supported.",
+                    )
+                bit = Qubit()
+                symbol = Symbol(name, bit, types.Qubit(), Scope.GLOBAL, node)
+                context.circuit.add_bits([bit])
+                self.symbols[name] = symbol
+            else:
+                raise_from_node(node, f"name '{name}' is not defined in this scope")
```
It's right to do the `is_physical` check here, but after that, we should be attaching it as information in the type system, rather than requiring other places to do the same semantic analysis later.

jakelishman commented on src/qiskit_qasm3_import/expression.py:
```diff
@@ -286,21 +317,27 @@ def visit_IndexedIdentifier(self, node: ast.IndexedIdentifier):
 
 
 def resolve_condition(
-    node: ast.Expression, symbols: Mapping[str, Symbol]
+    node: ast.Expression, context=None  # symbols: Mapping[str, Symbol]
 ) -> Union[Tuple[Clbit, bool], Tuple[Iterable[Clbit], int]]:
     """A resolver for conditions that can be converted into Qiskit's very basic equality form
     of either ``Clbit == bool`` or ``ClassicalRegister == int``.
 
     This effectively just handles very special outer cases, then delegates the rest of the work to a
     :class:`.ValueResolver`."""
 
+    if isinstance(context, dict):
+        symbols = context
+        context = None
+    else:
+        symbols = context.symbol_table
+
```
Similar to before: I don't think this should be optional, and if we're passing the whole thing, we don't need to separate `symbols` from the `State`.

jlapeyre replied:
I did this early, did not want to break any possible existing calls. Of course in the end, it's not needed.

jakelishman commented on tests/test_convert.py:
```diff
@@ -76,6 +76,69 @@ def test_readme_circuit():
     parse(source)
 
 
+def test_readme_circuit_physical_qubits():
+    source = """
```
I'm ok to add a second integration test based on the physical-qubit mode, but let's get rid of the lines that don't make sense rather than commenting them out.

jlapeyre replied:
I was wondering whether to fix them up or put in something else. But I prefer removing the cruft now.

jakelishman commented on tests/test_convert.py:
```diff
@@ -840,3 +926,28 @@ def test_alias_rejects_bad_types():
         parse(source)
+
+
+def test_reject_mixed_addressing_mode():
+    source1 = """
+    qubit q1;
+    reset $0;
+    """
+    with pytest.raises(
+        ConversionError, match="Physical qubit referenced in virtual addressing mode"
+    ):
+        parse(source1)
+
+    source2 = """
+    reset $0;
+    qubit q1;
+    """
+    with pytest.raises(ConversionError, match="Virtual qubit declared in physical addressing mode"):
+        parse(source2)
+
+    source3 = """
+    reset $0;
+    qubit[3] q;
+    """
+    with pytest.raises(ConversionError, match="Virtual qubit declared in physical addressing mode"):
+        parse(source3)
```
I'd marginally prefer to separate out the tests, so they all get tested individually rather than failing at the first.

Please can we also add some failure (and success) tests that we get suitable behaviour if the hardware qubits are only used within nested scopes, etc.

Hardware qubits should also be rejected from within `gate` bodies, regardless of the mode - gates can't reference them:

> The statements in the body can only refer to the symbols given in the parameter or argument list, and these symbols are scoped only to the subroutine body.

`gate` declarations should still work and be usable in hardware-qubit mode.

jakelishman reviewed (commented):
Thanks for looking at this.  I think this is a solid start.  I think we'd be better leveraging the type system and pushing a lot of this work onto the type checker, rather than doing it ad-hoc throughout the higher-level components, but that's not too difficult a change.  I also had a couple of data-representation points, but nothing super serious.

jlapeyre commented on src/qiskit_qasm3_import/data.py:
```diff
@@ -21,6 +21,35 @@ class Scope(enum.Enum):
     NONE = enum.auto()
 
 
+class AddressingMode:
+    """Addressing mode for qubits in OpenQASM 3 programs.
+
+    This class is useful as long as we allow only physical or virtual addressing modes, but
+    not mixed. If the latter is supported in the future, this class will be modified or removed.
+    """
+
+    # 0 == UNKNOWN
+    # 1 == PHYSICAL
+    # 2 == VIRTUAL
+
+    def __init__(self):
+        self._state = 0
+
+    def set_physical_mode(self):
+        """Set the addressing mode to physical. On success return `True`, otherwise `False`."""
+        if self._state != 2:
+            self._state = 1
+            return True
+        return False
+
+    def set_virtual_mode(self):
+        """Set the addressing mode to virtual. On success return `True`, otherwise `False`."""
+        if self._state != 1:
+            self._state = 2
+            return True
+        return False
```
The commits are squashed in this PR. But I did use `Enum` earlier: https://github.com/jlapeyre/qiskit-qasm3-import/commit/a4491f4bce9425697f4f4f05d6f499783a0e7ea6 . I didn't like that the details of how we switch modes and what is allowed, and how this is implemented, are spread across a few files and sections of code. I wanted it in one place and behind an abstraction. I noted in the commit message for the current code, https://github.com/jlapeyre/qiskit-qasm3-import/commit/0834755ea5419b0539ceca24349430c14132886c, that the abstraction leaks. But if you compare the code you may agree that it is better this way than with `Enum`. I considered raising the error in `AddressingMode`, which would hide the abstraction at the cost of another call in the call stack. I think that would be a good trade.

I agree with your argument that this does not belong in `data.py`.

jakelishman replied:
Personally I think that `Enum` code looks much cleaner, and because it doesn't try to abstract very simple mutation away, it doesn't leak either. If you really did encapsulate the setter and erroring out, I'd use an enum and put the setter logic in `State`, but really I'd prefer to just revert exactly to that commit you had.

jlapeyre replied:
Let me take another try at a non-leaky abstraction. If it doesn't look good we can revert to Enum. My thought is that with a good abstraction  you have one instantiation of an `AddressingMode` (with no explicit initial data passed). And two calls in two files, one to set physical addressing, the other to set virtual. If there is a problem you get an error.

Compare to `Enum` here: https://github.com/jlapeyre/qiskit-qasm3-import/commit/a4491f4bce9425697f4f4f05d6f499783a0e7ea6 . First, `AddressingMode` should be, as you say, moved, to `converter.py`. But it must be initialized with a value `UNKNOWN`. Then in two different files there are two reads of the data and two mutations. If I have a good abstraction, then I change the implementation, or with luck even the semantics, without changing code at the call sites. But with `Enum` I have to find and understand all five uses of the data and the logic surrounding them.

jlapeyre commented on src/qiskit_qasm3_import/converter.py:
```diff
@@ -183,13 +191,23 @@ class ConvertVisitor(QASMVisitor[State]):
     # sanity, as the reference package doesn't yet do this.
 
-    # pylint: disable=missing-function-docstring,no-self-use,unused-argument
+    # pylint: disable=missing-function-docstring,no-self-use,unused-argument,protected-access
```
It's in line 209: `state.circuit._layout = TranspileLayout(initial_layout, input_qubit_mapping)`. The property `_layout` doesn't seem to be  described anywhere and there doesn't appear to be an API.

jakelishman replied:
Ah fair. It'll become public in Terra 0.24 probably, but yeah, for now let's just move that particular suppression to be much more locally scoped.

jlapeyre commented:
Because there were several items requested in the review, I added another checklist in the opening comment. All items except adding some more tests have been addressed. 

Handling of the addressing mode is encapsulated in a class. This is slightly more complex structure-wise than using just an `Enum`, but not by much. It has the advantage that mutable details of the state of addressing mode are not spread across classes in two files. In general, it seems Qiskit prefers spreading mutation of state throughout the code. I suppose it is simpler in a sense and often is not problematic.  In the present case in particular, reading and writing the `Enum` instead of using the present class is not really too complex; this code is not very big. I'm ok with using the `Enum` instead.

There are probably a few commented out print statements left to remove.

jlapeyre commented:
Hardware qubits seen for the first time in local scope do not appear in the global symbol table. Local symbols are, correctly, discarded. I think this and other clumsy manipulations might be ameliorated by storing hardware qubits in a separate structure.

jakelishman commented:
But that's because you're adding them to the local scope and not the global scope.  They should be added to the global scope. That's not a clumsy manipulation, it's just following the rules of the language.  Having multiple symbol tables _is_ clumsy, and going to be a massive source of edge-case bugs if every lookup has to look in more than one place.

jakelishman commented on src/qiskit_qasm3_import/converter.py:
```diff
@@ -88,15 +93,45 @@
 
 
+class AddressingMode:
+    """Addressing mode for qubits in OpenQASM 3 programs.
+
+    This class is useful as long as we allow only physical or virtual addressing modes, but
+    not mixed. If the latter is supported in the future, this class will be modified or removed.
+    """
+
+    def __init__(self):
+        self._state = 0  # 0 == UNKNOWN, 1 == PHYSICAL, 2 == VIRTUAL
+
+    def set_physical_mode(self, node):
+        """Set the addressing mode to physical. On success return `True`, otherwise `False`."""
+        if self._state == 2:
+            raise_from_node(
+                node,
+                "Physical qubit referenced in virtual addressing mode. Mixing modes not currently supported.",
+            )
+        self._state = 1
+
+    def set_virtual_mode(self, node):
+        """Set the addressing mode to virtual. On success return `True`, otherwise `False`."""
+        if self._state == 1:
+            raise_from_node(
+                node,
+                "Virtual qubit declared in physical addressing mode. Mixing modes not currently supported.",
+            )
+        self._state = 2
+
+
 class State:
-    __slots__ = ("scope", "source", "circuit", "symbol_table", "_unique")
+    __slots__ = ("scope", "source", "circuit", "symbol_table", "_unique", "addressing_mode")
```
The addressing mode is a property of the _overall_ state of the parser, not just the state within a frame on the evaluation stack.  The addressing mode is purely a global-state thing, so the current mode must be accessible to `ConversionVisitor` at the point it's popping things off the stack.  I don't think this current form with local encapsulation will work with hardware qubits first encountered within a local scope.

I still don't like the class-based encapsulation, but I will go along with some form of it around the mode if you really prefer it.  However, I will insist that we use an `enum` for the mode - a comment explaining the values is not a substitute for correct typing.

jlapeyre replied:
> The addressing mode is purely a global-state thing

Yep this is clear.

> I don't think this current form with local encapsulation will work with hardware qubits

I don't think so either. I didn't worry about whether to copy or reference the addressing mode in the construction of a the `State` for local scope. Because either way its broken. 

> That's not a clumsy manipulation, it's just following the rules of the language.

Yes, I realize that local scope and global scope are not the same thing. I wasn't clear about what I think is clumsy. The example I had in mind is that we search the entire symbol table at the end to see if there is a hardware qubit in it. I think the design needs to change a bit. Two symbol tables is just one idea. I did check how many places would have to search both of them, and it looked ok. But even storing the hardware qubits in a separate structure in `State` is not enough.

I considered an enum for the number labeling the mode as well, probably even prefer it. I omitted it because with a class wrapping the enum the code might start to look wordy. Plus the magic numbers are restricted to this one place in the code, which makes them less of an issue.

jakelishman commented on src/qiskit_qasm3_import/expression.py:
```diff
@@ -17,19 +17,33 @@
+    return re.match(_PHYSICAL_QUBIT_RE, name) is not None
+
+
+# Qiskit represents physical qubits in a layout by integers.
+# def physical_qubit_identifiers_to_ints(names):
+#     """Convert an iterable of identifiers of physical qubits to a list of corresponding `int`s."""
+#     return [int(name[1:]) for name in names]
```
Please can you run through and check for dead code like this?

jakelishman commented on src/qiskit_qasm3_import/expression.py:
```diff
@@ -55,32 +69,47 @@ class ValueResolver(QASMVisitor):
+                    )
+                self.context.addressing_mode.set_physical_mode(node)
+                self.context.circuit.add_bits([bit := Qubit()])
+                symbol = Symbol(name, bit, types.HardwareQubit(), Scope.GLOBAL, node)
+                self.context.symbol_table[name] = symbol
```
This needs to be added to the global scope.  We probably want to more properly encapsulate the symbol table into a class that can hold its own stack at this point, and have `SymbolTable.get` and `SymbolTable.set_local` and `SymbolTable.set_global` methods on it.

jlapeyre replied:
I favor encapsulating the symbol table.

All of the state is in `State`. A new `State` is created for locate and gate scopes  (`FUNCTION` and `CALIBRATION` are currently unused). These structures are temporary and no information is propagated explicitly. Rather it is controlled by either copying or referring to the data in the parent `State`. A global symbol table in `State` might work with this scheme. I didn't think through the details. But encapsulated access to symbol tables might take care of the complexity of handling two tables. For example you can query a symbol name, or restrict to only local or global symbols.

jlapeyre replied:
I am resolving this conversation. The design of `SymbolTable` is discussed in another conversation in this review.

jakelishman commented on src/qiskit_qasm3_import/expression.py:
```diff
@@ -286,15 +315,15 @@ def visit_IndexedIdentifier(self, node: ast.IndexedIdentifier):
 
 def resolve_condition(
-    node: ast.Expression, symbols: Mapping[str, Symbol]
+    node: ast.Expression, context: "State"
```
Not sure why the quotes are necessary here - `State` is defined.

jlapeyre replied:
Last time I tried, I got a "partially initialized module" (or something like that) error.

jlapeyre replied:
Specifically,  removing the quotes and  importing `State` as in this [modification of expression.py](https://github.com/jlapeyre/qiskit-qasm3-import/blob/remove-quotes-from-state/src/qiskit_qasm3_import/expression.py) gives
```
ImportError: cannot import name 'State' from partially initialized module 'qiskit_qasm3_import.converter' (most likely due to a circular import) (/home/lapeyre/qiskit/qiskit-qasm3-import/src/qiskit_qasm3_import/converter.py)```

There may be a way around this. I did not yet try.

jlapeyre replied:
Moving `State` to a new file `state.py` is a possibility. I made this change in [another branch](https://github.com/jlapeyre/qiskit-qasm3-import/tree/remove-quotes-from-state/src/qiskit_qasm3_import) and all tests pass.

jlapeyre replied:
In an earlier commit, I introduced `state.py` to this PR. This takes care of the problem here. Resolving converstation.

jakelishman commented on src/qiskit_qasm3_import/types.py:
```diff
@@ -109,6 +109,16 @@ def pretty(self):
         return "qubit"
 
 
+@typing.final
+class HardwareQubit(Type):
+    """A hardware qubit.  This corresponds a hardware qubits referenced in Terra's :class:`~qiskit.transpiler.TranspilerLayout`."""
+
+    __slots__ = ()
+
+    def pretty(self):
+        return "hardware qubit"
```
This needs adding to the docs in `docs/index.rst` as well.

jakelishman commented on tests/test_convert.py:
```diff
@@ -10,10 +10,14 @@
 
 from qiskit_qasm3_import import parse, ConversionError
 
+# pylint: disable=too-many-lines
+
```
Let's disable this project-wide in `pyproject.toml` - it's a stupid rule anyway.

jlapeyre commented:
I'm experimenting with implementing a class `SymbolTable` rather than using a simple `dict`. This does
fix the one test for broken behavior. (i.e. the test for broken behavior now fails because the tested behavior is correct)
However, other new bugs still exist.

jlapeyre commented:
@jakelishman what does "per-run state" mean here?
https://github.com/Qiskit/qiskit-qasm3-import/blob/246e999573c0d7a2162b779d9525b8b5b79c3cf0/src/qiskit_qasm3_import/expression.py#L56

I am updating doc strings. The doc string in question says that the instance state of `ValueResolver` is only symbols, which is no longer true.

jakelishman commented on tests/test_convert.py:
```diff
@@ -840,3 +973,154 @@ def test_alias_rejects_bad_types():
+    assert qc._layout == layout
+
+
+def test_hardware_qubit_first_seen_in_local_scope():
+    source = """
+    include "stdgates.inc";
+
+    bit[2] mid;
+
+    while (mid == "00") {
+        h $100;
+        h $101;
+        mid[0] = measure $100;
+        mid[1] = measure $101;
+     }
+    """
+
+    qc = parse(source)
+    layout = _make_layout([100, 101])
+    assert qc._layout == layout
```
I think we also need to test that these hardware qubits can then be used outside the local scope they're first seen in, and that that produces expected outputs (i.e. still refers to the same qubit) and ideally the lookup doesn't involve any copies.

As a corollary of that: can we just that the addressing mode getting fixed to "hardware" within a nested scope causes the expected failures if a virtual qubit is first declared _after_ (e.g. the program `bit[2] mid; while (mid == "00") { h $0; } qubit q;` should raise an error).

jlapeyre replied:
I agree these are good tests to have. I'm not sure what you mean by "lookup doesn't involve any copies", and furthermore what you'd like to test about it. I'll add this test first. Then implement your alternative design above for `SymbolTable`.

jakelishman replied:
I meant something along the lines that "seeing `$0` in a new scope doesn't cause a new entry to be added to the symbol table; it should already be there", but I didn't explain it very well.  It's not actually super important to try and test that separately, I think - it's kind of an implementation detail.

jlapeyre replied:
Test exception raised if: Fix HW addressing mode in local scope. Then declare virtual qubit.
Done in efa1e031ab

jlapeyre replied:
Test hardware qubit use after introduction in a previous local scope
Done in 855dc138303

jakelishman commented on tests/test_values.py:
```diff
@@ -22,52 +30,63 @@
+        q0, q0_type = resolver.resolve(ast.Identifier(name=name))
+        assert isinstance(q0, Qubit)
+        assert isinstance(q0_type, types.HardwareQubit)
+    with pytest.raises(ConversionError, match="not defined in this scope"):
+        resolver.resolve(ast.Identifier(name="$q"))
```
This isn't a valid OQ3 identifier, so we probably don't need to make any particular assertions about what happens with it.

jlapeyre replied:
I agree "not defined in scope" is not correct, but I often like to test the behavior of the code as it is, even though it might be improved. (e.g. test that wrong results happen, or that a test is broken in the expected way). ... But, I agree it seems to be testing something that's outside the scope of this converter. I don't recall why I put it here.

jakelishman replied:
I'm fine to leave it.

`qss-qasm` accepts `$q` because historically that was allowed as a "wildcard" identifier for a hardware qubit in a `defcal` statement, but that was changed in openqasm/openqasm#371.

jlapeyre replied:
Remove test involving invalid identifier `$q` done in 94fb45f50446

jlapeyre replied:
> I'm fine to leave it.
I removed it before I saw this. Maybe I didn't refresh. But I'm ok with removing it.

jakelishman commented on src/qiskit_qasm3_import/state.py:
```diff
@@ -0,0 +1,233 @@
+            if symbol := table.get(name):
+                return symbol
+        return None
+
+    def hardware_qubits(self):
+        """Return an iterator over the `Symbol`s referring to hardware qubits."""
+        return (
+            sym for sym in self._global_table.values() if isinstance(sym.type, types.HardwareQubit)
+        )
```
This feels weirdly over-specific.  I'd prefer to either let/make consumers of the symbol table do the iteration themselves, or to define a general `filter` iterator that takes an arbitrary callable.

jlapeyre replied:
I see your point. There are other reasons to not like this, eg, this is only used once.

jakelishman reviewed (commented):
This is looking much better, thanks.  Most of my comments are minor nitpicks, but I do have some serious stuff about the implementation of the `SymbolTable` and some missing tests.

I think it's important to test that the `AddressingMode` is being set correctly and propagated all the way up to the global scope.  As I'd understood it, that was the reason why you wanted to have the specific `AddressingMode` interior-mutability class, but as far as I can tell, this current implementation doesn't _share_ that mutability between different levels of `State` in the (implied) stack to propagate the change.  I suggested a test to add to check that we are satisfying that.

Can we please also add a test that we can use hardware addressing in the main program, but still have user-defined gates?

For layout: I think you might need to check me, but I'm fairly sure Qiskit says that backends are required to label their hardware qubits starting from 0 with no gaps, so I think our outputs will need to include those implicit qubits as well.  I think we probably need to add an option to the top-level `convert` function to allow the user to specify the total size of the backend as well, so for example if they give us a circuit that contains only `$5`, they can tell us that their backend actually has 10 qubits so we can output the correct full width.

jakelishman commented:
Oh, and to answer that question: I _think_ I meant that `ValueResolver` doesn't have any mutable state that it generated/tracked as it traversed the expression tree, it just had a static lookup.  That text isn't clear though, and I don't entirely remember what I was thinking.

jlapeyre commented on src/qiskit_qasm3_import/converter.py:
```diff
@@ -25,114 +23,18 @@
 from .exceptions import ConversionError, raise_from_node
-from .expression import ValueResolver, resolve_condition
-
-
-_STDGATES = {
```
That was inadvertent. I agree it belongs in `converter.py` as it is only used there.

jakelishman replied:
I think the PR might be out-of-sync with the comment - this has still been moved out of `converter.py` in the currently pushed iteration.

jlapeyre commented on src/qiskit_qasm3_import/converter.py:
```diff
@@ -25,114 +23,18 @@
-        out.symbol_table = self.symbol_table.copy()
-        out._unique = self._unique
-        return out
+from .expression import (
+    ValueResolver,
+    resolve_condition,
+)
```
What you have looks better if the linter an black accept it. I guess maybe black made the awkward change, and I subsequently edited it or removed something.

jakelishman replied:
It's just because of the trailing comma - Black puts them all on separate lines if there is one, and collapses them (if they fit) if there isn't.  Not a big deal.

jlapeyre commented on src/qiskit_qasm3_import/expression.py:
```diff
@@ -52,35 +61,51 @@ class ValueResolver(QASMVisitor):
+                        f"Illegal qubit reference '{name}'. References to hardware qubits not allowed in gate definitions.",
+                    )
+                self.context.addressing_mode.set_physical_mode(node)
+                self.context.circuit.add_bits([bit := Qubit()])
+                symbol = Symbol(name, bit, types.HardwareQubit(), Scope.GLOBAL, node)
```
Agreed. "new scope": Yes that acts a bit like a `let` construct. But this case `bit := Qubit...` is different. It can go.

jlapeyre replied:
Done in 920354519340

jlapeyre commented on src/qiskit_qasm3_import/expression.py:
```diff
@@ -52,35 +61,51 @@ class ValueResolver(QASMVisitor):
-        symbol = self.symbols[name]
+        if name not in self.context.symbol_table:
+            if is_physical(name):  # Physical qubits are not declared.
+                if self.context.scope == Scope.GATE:
```
Yeah, IIRC it does complain about `== None`.

jlapeyre replied:
> `Enum` instances can (should) be compared by is checks

Done in 07480352143

jlapeyre replied:
Found another one and fixed it in 1d8b5f33c68b. Resolving

jlapeyre commented on src/qiskit_qasm3_import/state.py:
```diff
@@ -0,0 +1,233 @@
+
+    def insert(self, symbol: Union[List, Symbol]):
+        """Insert a `Symbol` (or each of a list thereof) into the symbol table."""
+        if isinstance(symbol, list):
+            for sym in symbol:
+                self._insert(sym)
+        else:
+            self._insert(symbol)
```
I don't know the most satisfactory way to handle this in Python. I wanted to avoid the following, which I often see in Python and Qiskit  in particular
```python
def insert(self, symbol: Union[List, Symbol]):
    if not isinstance(symbol, list):
        symbol = [symbol]
    for sym in symbol:
        if symbol.scope == Scope.GLOBAL:
            self._global_table[symbol.name] = symbol
        else:
            self._local_table[symbol.name] = symbol
```
You can patch this up to detect tuples, etc.

But in fact I think I just have `insert` support single symbols. All calls that send more than one symbol come through a single function. I pretty sure I can put the loop there (`tests/test_values.py` , `_make_context`).

jakelishman replied:
Yeah, I meant I expect the caller to handle simple loops like this (e.g. `for symbols in symbols: table.insert(symbol)`).  There's a reason to do autovectorisation/broadcasting in something like Numpy, but not in raw Python imo - having the explicit loops are clearer at the call site, which is important for code clarity.

If you were going to do the autobroadcast, I'm not sure why you want to avoid the form you've given there - the code reuse is much simpler without needing an inner function (except in this case it should be `if isinstance(symbol, Symbol): symbol = [symbol]` so it worked with arbitrary iterables).

jlapeyre commented on src/qiskit_qasm3_import/converter.py:
```diff
@@ -186,10 +88,25 @@ class ConvertVisitor(QASMVisitor[State]):
-        return self.visit(node, State(Scope.GLOBAL, source)).circuit
+        """Convert a program node into a :class:`~qiskit.circuit.QuantumCircuit`.
+
+        If given, `source` is a string containing the OpenQASM 3 source code that was parsed into `node`.
+        This is used to generated improved error messages. A :class:`.State` containing
+        information about the conversion is returned. The :class:`~qiskit.circuit.QuantumCircuit` is
+        stored in property thereof named `circuit`.
```
I see that some comments also exceed 100 characters. Unlike docstrings, these have no effect on the user. But we I could wrap these. For example
https://github.com/Qiskit/qiskit-qasm3-import/blob/246e999573c0d7a2162b779d9525b8b5b79c3cf0/src/qiskit_qasm3_import/converter.py#L66

jlapeyre commented on src/qiskit_qasm3_import/converter.py:
```diff
@@ -288,8 +205,8 @@ def _resolve_constant_duration(self, node: ast.Expression, context: State) -> Tu
-        if not isinstance(type, (types.Int, types.Uint, types.Angle, types.Float)):
+        value, _type = self._resolve_generic(node, context)
+        if not isinstance(_type, (types.Int, types.Uint, types.Angle, types.Float)):
             raise_from_node(node, "required an angle-like value")
```
Renamed in ef0b676c3

jlapeyre commented on src/qiskit_qasm3_import/expression.py:
```diff
@@ -52,35 +61,51 @@ class ValueResolver(QASMVisitor):
-    def generic_visit(self, node: ast.QASMNode, context: None = None):
+    # pylint: disable=arguments-differ
+    def generic_visit(self, node: ast.QASMNode):
         raise_from_node(node, f"'{node.__class__.__name__}' cannot be resolved into a Qiskit value")
```
Done in 9810016f27e5

jlapeyre commented on src/qiskit_qasm3_import/state.py:
```diff
@@ -0,0 +1,233 @@
+        out._global_table = self._global_table
+        out._builtin_table = self._builtin_table
+
+        return out
```
I agree this design is more appealing. This has a more stack-like structure. And so do scopes. And it's probably more efficient because it avoids copying. It might be more efficient when iterating over symbols to separate some kinds of symbols, e.g. hardware qubits. But thats an optimization.

Also, you inadvertently omitted looking for symbols in `global_symbols`, right?

jakelishman replied:
I don't think I omitted any lookup here?  The table first tries to look in itself (the local scope).  If that fails and there's no parent scope, it complains "symbol not found".  Otherwise it asks its parent scope to try a lookup.  Once it's got a match it finally checks to see if it's allowed to "see" that symbol.  The recursive `self.base` lookup should handle going all the way up to the global scope?

I don't think it's a good idea to try and separate out various symbols without a good cause. It opens more surface area for bugs (you need to check more places for definitions / keep more data in sync), and the number of symbols we're going to have to search through really isn't going to be all that high.

jlapeyre replied:
In any case, at least for the first iteration, I think it's best to have a simple implementation. It doesn't look like it will be wasted effort if we need to make it more complicated later.

jlapeyre commented on src/qiskit_qasm3_import/expression.py:
```diff
@@ -52,35 +61,51 @@ class ValueResolver(QASMVisitor):
-        if name not in self.symbols:
-            raise_from_node(node, f"name '{name}' is not defined in this scope")
-        symbol = self.symbols[name]
+        if name not in self.context.symbol_table:
```
Fixed in 7606c704b1c6c

jlapeyre commented:
@jakelishman I think that moving logic for visibility into the `SymbolTable`  (given that `SymbolTable` has a nice stack-like design for handling scopes) is the right thing for several reasons.  But in raising exceptions we want to include information on the `node`; in particular the location in the source file. I want to just pass this information, for example the call would be `symbol_table.get(symbol_name, node)`. This doesn't feel right because node and source information make the interface less clean. They are only for diagnostics.  The same issue arose with `AddressingMode` and I chose the same solution; passing `node`.

Regarding the symbol table. We could do one of the following
1. Not pass `node` so that we lose the information for diagnostics. This results in a cleaner, less complex interface.
2. Pass `node` wherever it is needed for diagnostics. This means, even though symbol_table has a nice map-like interface to hide complexity, a simple lookup requires passing extraneous information (that will never be  used for a correct input program)
3. Note the the context and node are almost always passed together. On recursive calls a different node is passed and the same context (possibly modified) is passed. Passing a light-weight structure, perhaps a named tuple, with these two pieces of information would hide some complexity. The arguments for this decreasing bugs are the standard ones, which I have already run into here: argument order, typos, etc.  You have more complexity in one sense, but surface area for bugs is decreased. Same argument for why I like `AddressingMode` to be as struct.

At present, I think choosing 2. above is best. Get the kinks worked out, and then evaluate whether progressing to 3. is a good idea.

jakelishman commented:
Yeah, I agree - it's not a perfect situation, but I don't think your number 2 is too tricky an interface, and still lets us emit diagnostics (at least the very limited set that we do).

jlapeyre commented:
Yes, number two is the easiest way forward and doesnt sacrifice the user facing behavior that we want.

jlapeyre commented:
@jakelishman I pushed more changes last night. Mainly this is using the new implementation of `SymbolTable`. Identifying errors in the test suite by checking strings was getting a bit complex. In the case of hardware qubits in a gate definiton, the error is raised in two places and would require keeping the strings in sync. I instead raise an error `HardwareQubitInGateError`.


I have yet to change the mapping of virtual to physical qubits. What I have now is in any case incorrect, because it ignores `Qubit` objects in the circuit and creates a new register. Further processing to take this to a backend would fail.

jlapeyre commented:
> I think we don't in fact need to worry about the number of hardware qubits ...

https://github.com/Qiskit/qiskit-terra/issues/9609

We do need to do something to better support the problem in the issue above.

jakelishman commented on src/qiskit_qasm3_import/exceptions.py:
```diff
@@ -8,9 +8,23 @@ class ConversionError(Exception):
+    raise ConversionError(message, node)
+
+
+class PhysicalQubitInGateError(ConversionError):
+    def __init__(self, name: str, node=None):
+        self.name = name
+        msg = (
+            f"Illegal qubit reference '{name}'. References to hardware "
+            "qubits not allowed in gate definitions."
+        )
+        super().__init__(msg, node)
```
I'm not a fan of doing this in a separate class, mostly because in an ideal world we wouldn't be raising exceptions at all, we'd be appending the messages to a diagnostic stream and then only finally raising an exception (library) / returning a non-zero error code (application).  There's also the weird API where a user could ask "why is _this_ error separate and not _that_ error?"; it may well be subjective, but imo we should only use separate exception classes when we actively expect users to want to behave differently upon `try/except`'ing them.

I think you can achieve what you want (unified error message) without the new class if you use a little helper function to encapsulate the string.

jlapeyre replied:
Using a separate method only for this one case is only because I didn't want to take the time to redesign everything.

I agree 100% that exceptions are not a priori the correct way to handle these errors in translation. In practice, doing something other than exiting immediately would be more complex.

I did consider using a scheme other than exceptions to make this more clean; some kind of tag. But the exception machinery is tempting because most of the work is already done for us. I'll try something else.

jakelishman commented on src/qiskit_qasm3_import/expression.py:
```diff
@@ -17,19 +17,37 @@
+    return re.match(_PHYSICAL_QUBIT_RE, name) is not None
+
+
+# A hardware-qubit symbol has the form '$' followed by digits.
+# The digits are the identifier used by backends.
+def hardware_qubit_map(symbol_table: SymbolTable):
+    "Return a `dict` mapping `Qubit` instances to `int`s representing physical qubit identifiers."
+    return {sym.data: int(sym.name[1:]) for sym in symbol_table.globals() if is_physical(sym)}
```
Probably nit-picky, but since this is only used in `converter.py`, I think it logically belongs there?

I do still think that the number of hardware qubits is important for useful reconstruction.  We might be able to make a valid `Layout` object with holes in it, but in practice, that doesn't fully represent a hardware-mapped circuit, because those `QuantumCircuit` objects always have as many qubits in them as the backend has, where the unused qubits are virtually mapped to a (non-existent) quantum register called `ancilla`:
```python
In [1]: from qiskit import QuantumCircuit, transpile
   ...: from qiskit.providers.fake_provider import FakeKolkata
   ...:
   ...: qc = transpile(QuantumCircuit(2), FakeKolkata())
   ...: print(qc.num_qubits)
   ...: print(qc._layout.initial_layout)
27
Layout({
0: Qubit(QuantumRegister(2, 'q'), 0),
1: Qubit(QuantumRegister(2, 'q'), 1),
2: Qubit(QuantumRegister(25, 'ancilla'), 0),
3: Qubit(QuantumRegister(25, 'ancilla'), 1),
4: Qubit(QuantumRegister(25, 'ancilla'), 2),
5: Qubit(QuantumRegister(25, 'ancilla'), 3),
6: Qubit(QuantumRegister(25, 'ancilla'), 4),
7: Qubit(QuantumRegister(25, 'ancilla'), 5),
8: Qubit(QuantumRegister(25, 'ancilla'), 6),
9: Qubit(QuantumRegister(25, 'ancilla'), 7),
10: Qubit(QuantumRegister(25, 'ancilla'), 8),
11: Qubit(QuantumRegister(25, 'ancilla'), 9),
12: Qubit(QuantumRegister(25, 'ancilla'), 10),
13: Qubit(QuantumRegister(25, 'ancilla'), 11),
14: Qubit(QuantumRegister(25, 'ancilla'), 12),
15: Qubit(QuantumRegister(25, 'ancilla'), 13),
16: Qubit(QuantumRegister(25, 'ancilla'), 14),
17: Qubit(QuantumRegister(25, 'ancilla'), 15),
18: Qubit(QuantumRegister(25, 'ancilla'), 16),
19: Qubit(QuantumRegister(25, 'ancilla'), 17),
20: Qubit(QuantumRegister(25, 'ancilla'), 18),
21: Qubit(QuantumRegister(25, 'ancilla'), 19),
22: Qubit(QuantumRegister(25, 'ancilla'), 20),
23: Qubit(QuantumRegister(25, 'ancilla'), 21),
24: Qubit(QuantumRegister(25, 'ancilla'), 22),
25: Qubit(QuantumRegister(25, 'ancilla'), 23),
26: Qubit(QuantumRegister(25, 'ancilla'), 24)
})
```

I view this support of hardware qubits as moving us towards being able to replace the direct-string OQ3 input form for `dynamic=True` backends in IBM provider, which is currently passed directly to the QSS stack.  To be as close to a drop-in replacement for that as possible, we'll need to make it easy to produce a `QuantumCircuit` object that could be the direct output of `transpile`, which in this case means being full-width on the backend.

jlapeyre replied:
> Probably nit-picky, but since this is only used in converter.py, I think it logically belongs there?

I'm fine with moving it to converter.py. This will require importing `SymbolTable` and `is_physical` into converter.py

jlapeyre replied:
Fixed in 9199a6a . Also, after removing spurious `assert`s, I am able to remove `SymbolTable` from imports into expression.py

jlapeyre replied:
Fixed `_STDGATES` location in bf80706

jakelishman commented on src/qiskit_qasm3_import/expression.py:
```diff
@@ -77,10 +97,19 @@ def generic_visit(self, node: ast.QASMNode, context: None = None):
-            raise_from_node(node, f"name '{name}' is not defined in this scope")
-        symbol = self.symbols[name]
+        cxt = self.context
+        assert isinstance(cxt.symbol_table, SymbolTable)
```
Is this `assert` left over from debugging?

jlapeyre replied:
Fixed in 4a19f39

jakelishman commented on src/qiskit_qasm3_import/expression.py:
```diff
@@ -77,10 +97,19 @@ def generic_visit(self, node: ast.QASMNode, context: None = None):
-        symbol = self.symbols[name]
+        cxt = self.context
+        assert isinstance(cxt.symbol_table, SymbolTable)
+        if (symbol := cxt.symbol_table.get(node.name, node)) is not None:
+            return symbol.data, symbol.type
```
Nice walrussing - I'm a fan of this sort of use of it.

jlapeyre replied:
What we really need is [anaphoric macros](https://en.wikipedia.org/wiki/Anaphoric_macro). ( ha ;) )

jakelishman commented on src/qiskit_qasm3_import/state.py:
```diff
@@ -0,0 +1,205 @@
+        if self.scope is Scope.GATE:
+            if (
+                isinstance(symbol.type, types.Gate)
+                or isinstance(
+                    symbol.type, (types.Int, types.Uint, types.Float, types.Angle, types.Duration)
+                )
+                and symbol.type.const
```
Super minor: can we parenthesis the `isinstance(symbol.type, (...)) and symbol.type.const` check?  I know it's valid by the operator precedence once I think about it, but personally it's not something that I can just "see" automatically in the same way as mathematical rules.

jlapeyre replied:
I agree. I copied it from here ( :laughing-and-crying-while-rolling-on-the-floor: )     
https://github.com/Qiskit/qiskit-qasm3-import/blob/246e999573c0d7a2162b779d9525b8b5b79c3cf0/src/qiskit_qasm3_import/converter.py#L107-L116

But seriously, it's not so minor. I think it will make it easier for the next person to read.

jlapeyre replied:
Fixed in e894ba1

jakelishman replied:
Ha, guess it was me who did it originally then!  Thanks for the change.

jakelishman commented on src/qiskit_qasm3_import/state.py:
```diff
@@ -0,0 +1,205 @@
+    def __init__(self, scope: Scope, base: Optional["SymbolTable"] = None):
+        self.symbols = {}
+        self.scope = scope
+        self.base = base
+        self.global_symbols = base.global_symbols if base is not None else _BUILTINS.copy()
```
I think it's better not to be accessing partial state of a `SymbolTable` (e.g. the backing dict for globals); it's likely to be a source of bugs if we need to do additional consistency management on `SymbolTable` operations, because the outer symbol table will end up being responsible for managing an secondary table's consistency, rather than just using the API to delegate.

If you're concerned about performance of fetching the `global_symbols`, then you could store a pointer to the global `SymbolTable` here, which is permitted to be the same pointer as `self.base`.

jlapeyre replied:
I don't follow. Could you clarify, please? I took the basic idea from [here](https://github.com/Qiskit/qiskit-qasm3-import/pull/2#discussion_r1104892623). The current implementation is slightly awkward: the root symbol table creates a `dict` for globals. Then children copy a reference to this table. At least this encapsulates the global table. An alternative would be to make a reference to a `dict` in the global scope `_GLOBAL_SYMBOLS = _BUILTIS.copy()`. Then each table would use `_GLOBAL_SYMBOLS` directly for access.

jakelishman replied:
I'm saying that in this current design, `global_symbols` is a pointer to the `symbols` dict of a different `SymbolTable` instance, rather than being a pointer to the entire class.  That means that any mutations to that dict are going to be fragile for the `SymbolTable` that owns it, if the `SymbolTable` class evolves, because we're directly mutating internal state of some object without using the encapsulating methods, and mutating this state affects many other classes.

I'm suggesting that `global_symbols` should just be the "top" `SymbolTable` of the `SymbolTable.base` tree, and not just a part of the internal state of a `SymbolTable`.  Within that, the two things I suggested were either dropping `self.global_symbols` as an attribute and replace it with a method:
```python
class SymbolTable:
    def global_symbols(self) -> Self:
        table = self
        while table.base is not None:
            table = table.base
        return table
```
or if you think that it's critically important for performance to keep the attribute cached, then:
```python
class SymbolTable:
    def __init__(self, base: Optional[Self] = None):
        self.base = base
        if base is None:
            self.symbols = _BUILTINS.copy()
            self.global_symbols = self
        else:
            self.symbols = {}
            self.global_symbols = self.base.global_symbols
```

jlapeyre replied:
I think I sorta follow what you mean.

This is not going to work:
```python
self.global_symbols = self
```
This `self.global_symbols = self.base.global_symbols` is similar to what I already had (not the same)--- a direct reference from leaves in the tree.

I think a good approach is to decide on an API that is insensitive to whether it's implemented more or less the way I already have it or something more like your idea. I'll try something....

jlapeyre replied:
The symbol table is completely redesigned in 11c718b

jakelishman commented on src/qiskit_qasm3_import/state.py:
```diff
@@ -0,0 +1,205 @@
+        target = self.global_symbols if symbol.type == types.HardwareQubit() else self.symbols
+        target[symbol.name] = symbol
+
+    def globals(self):
+        if self.global_symbols is None:
+            return []
+        return self.global_symbols.values()
```
In the current situation, I think `self.global_symbols` is always not `None`, though see above comment since I'm not sure it should be a raw dict.

jlapeyre replied:
Yes that's correct. This must be a relic.

jlapeyre replied:
If `global_symbols` is not a raw `dict` what should it be?

jlapeyre replied:
Fixed in 1cd0d3a . But we may redesign this anyway.

jakelishman commented on src/qiskit_qasm3_import/state.py:
```diff
@@ -0,0 +1,205 @@
+        # expect; the nodes have references to the complete source in their spans.
+        self.source = source
+        self.init_inner()
+
+    def init_inner(
```
Super minor, but probably let's make this private (`_init_inner`).  I don't think I was especially strict on that split when I wrote most of the package, but this one feels fairly cut-and-dried.

This DRY form looks much cleaner that what I'd originally written though, so thanks for that!

jlapeyre replied:
That's what I had originally, but removing the underscore was the fastest way to silence a spurious error: `W0212: Access to a protected member _init_inner of a client class`. Better to suppress the linter here.

jlapeyre replied:
Fixed in 33ee260

jakelishman replied:
Oh yeah, I see.  I agree the lint suppression is the right call there - it's not really a client, it's a part of the constructor method.

jakelishman commented on tests/test_conditions.py:
```diff
@@ -6,177 +6,200 @@
     return ast.BinaryExpression(ast.BinaryOperator["=="], left, right)
 
 
+def _make_context(symbols=None):
+    context = State(Scope.GLOBAL)
```
Pretty sure that this a bug; the only positional argument to `State` is the `source` string.

jlapeyre replied:
Yes. this is another relic.

jlapeyre replied:
Fixed in 68c6541

jakelishman reviewed (commented):
Ace, thanks John.  I have one small thing about the symbol table, one about the new exception, and I do think the "number of qubits in the backend" thing is something we need to take care of, but this is definitely coming in for landing.

jakelishman reviewed (commented):
I think if we could just make the slightly safer typing in the `SymbolTable`, we can merge this.  There's the question of "number of qubits in the backend", but let's do that in a follow-up.

jlapeyre commented:
> There's the question of "number of qubits in the backend", but let's do that in a follow-up.

A good idea. This PR discussion is already complicated.

jlapeyre commented on tests/test_convert.py:
```diff
@@ -235,6 +308,15 @@ def test_only_global_declarations():
+        include 'stdgates.inc';
+        qubit p;
+    """
+    with pytest.raises(ConversionError, match="already inserted in symbol table"):
```
This isn't really a good error message. If these were really errors,  we would trap it and throw  a more specific error. Because trying to insert a symbol twice may arise from different operations. But since we really want to handle the messages as diagnostics, not errors, we need another way to get the same effect.

jlapeyre commented:
@jakelishman I redesigned the symbol table. It is a stack rather than a recursive structure. I think this is much more in line with what you were thinking about above. Furthermore the stack is managed by context managers.

jlapeyre commented:
Fixes for a couple of outstanding issues: 

* I refactored until `PhysicalQubitInGate` was called in only one place. Then I replaced it with a call to `raise_from_node`. I check the identifier name (a `str`) rather than checking for `types.HardwareQubit` because I need to catch an illegal use where the hardware qubit's first appearance is in the gate, in which case it has not yet been constructed.

*  I refactored `_check_visible`.
